Add isBackwardCompatibleWith for Argument and FunctionSchema

houseroad opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23409 Add isBackwardCompatibleWith for Argument and FunctionSchema**

we intend to be conservative, and will relax the checks in future if necessary.
So far, we consider the following three conditions as backward compatible:
   1) two schemas are equal
   2) two schemas have same number of arguments, and this schema's
      arguments are backward compatible with the corresponding ones in
      argument list of old_schema.
   3) this schema has m argument, old_argument has n argument, m > n.
      the first n arguments of this schema are backward compatible with
      the corresponding arguments of old_schema. the remaning arguments
      must be either OptionalType or provide default values.

Differential Revision: [D16505203](https://our.internmc.facebook.com/intern/diff/D16505203/)

houseroad committed:
Add isBackwardCompatibleWith for Argument and FunctionSchema

we intend to be conservative, and will relax the checks in future if necessary.
So far, we consider the following three conditions as backward compatible:
   1) two schemas are equal
   2) two schemas have same number of arguments, and this schema's
      arguments are backward compatible with the corresponding ones in
      argument list of old_schema.
   3) this schema has m argument, old_argument has n argument, m > n.
      the first n arguments of this schema are backward compatible with
      the corresponding arguments of old_schema. the remaning arguments
      must be either OptionalType or provide default values.

Differential Revision: [D16505203](https://our.internmc.facebook.com/intern/diff/D16505203/)

houseroad committed:
Update on "Add isBackwardCompatibleWith for Argument and FunctionSchema"


we intend to be conservative, and will relax the checks in future if necessary.
So far, we consider the following three conditions as backward compatible:
   1) two schemas are equal
   2) two schemas have same number of arguments, and this schema's
      arguments are backward compatible with the corresponding ones in
      argument list of old_schema.
   3) this schema has m argument, old_argument has n argument, m > n.
      the first n arguments of this schema are backward compatible with
      the corresponding arguments of old_schema. the remaning arguments
      must be either OptionalType or provide default values.

Differential Revision: [D16505203](https://our.internmc.facebook.com/intern/diff/D16505203/)

dzhulgakov commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -78,6 +95,38 @@ struct Argument {
+  //   2) old_argument's type is T, and this Argument's type is Optional[T]
+  //   3) old_argument has no default value, and this Argument provides default
+  //      value
+  bool isBackwardCompatibleWith(const Argument& old_argument) const {
```
nit: there's function_schema_inl.h - maybe move implementation there (as it's lengthy)

dzhulgakov commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -78,6 +95,38 @@ struct Argument {
+    }
+    if (!(*type() == *old_argument.type() ||
+          (type()->isSubclass(TypeKind::OptionalType) &&
+           type()->cast<OptionalType>()->getElementType()))) {
```
you probably meant to compare getElementType() with old_argument's type

houseroad committed:
Update on "Add isBackwardCompatibleWith for Argument and FunctionSchema"


we intend to be conservative, and will relax the checks in future if necessary.
So far, we consider the following three conditions as backward compatible:
   1) two schemas are equal
   2) two schemas have same number of arguments, and this schema's
      arguments are backward compatible with the corresponding ones in
      argument list of old_schema.
   3) this schema has m argument, old_argument has n argument, m > n.
      the first n arguments of this schema are backward compatible with
      the corresponding arguments of old_schema. the remaning arguments
      must be either OptionalType or provide default values.

Differential Revision: [D16505203](https://our.internmc.facebook.com/intern/diff/D16505203/)

zdevito commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -78,6 +95,38 @@ struct Argument {
+      return false;
+    }
+    if (!(*type() == *old_argument.type() ||
+          (type()->isSubclass(TypeKind::OptionalType) &&
```
Please use `type()->cast<OptionalType>()` to test for it being optional, `isSubclass` is for internal use handling actual C++ subclassing.

houseroad replied:
I just saw `isSubtypeOfExt`, and used it.

zdevito commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -147,6 +188,49 @@ struct FunctionSchema {
+  //      the corresponding arguments of old_schema. the remaning arguments
+  //      must be either OptionalType or provide default values.
+  bool isBackwardCompatibleWith(const FunctionSchema& old_schema) const {
+    if (*this == old_schema) {
```
What is the purpose of these equality checks? I am wary because there should be no difference if these are deleted, but they _will_ make it harder to test whether the below code is correct.

houseroad replied:
removed

zdevito commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -147,6 +188,49 @@ struct FunctionSchema {
+      && overload_name() == old_schema.overload_name()
+      && is_vararg() == old_schema.is_vararg()
+      && is_varret() == old_schema.is_varret()
+      && returns() == old_schema.returns()
```
This is using Argument ==, is that was is intended rather than Arugment backward compatible?

houseroad replied:
originally, i just intended to be conservative. Now I changed the logical, use a inversed normal argument backcompat check.

zdevito commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -78,6 +95,38 @@ struct Argument {
+  // relax the checks in future. we consider the following the situations are
+  // backward compatible:
+  //   1) two arguments are equal
+  //   2) old_argument's type is T, and this Argument's type is Optional[T]
```
This is overly specific. Instead, it should be using the isSubtype property on on types, which already handles the optional case and will future proof it for further cases. Follow the subtyping rule for functions (https://en.wikipedia.org/wiki/Subtyping#Function_types)

houseroad replied:
agree, updated.

zdevito commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -147,6 +188,49 @@ struct FunctionSchema {
+  //   3) this schema has m argument, old_argument has n argument, m > n.
+  //      the first n arguments of this schema are backward compatible with
+  //      the corresponding arguments of old_schema. the remaning arguments
+  //      must be either OptionalType or provide default values.
```
OptionalType here would still be backward incompatible, because it would require that the old code pass None explicitly. The check should only be that more arguments have default values.

houseroad replied:
good catch, removed optionaltype condition.

zdevito commented on test/test_function_schema.py:
```diff
@@ -10,8 +11,81 @@ def test_serialize_and_deserialize(self):
+        new_schema = parse_schema('any(Tensor self, int dim, bool keepdim=False) -> Tensor')
+        self.assertTrue(new_schema.is_backcompat_with(old_schema))
+        self.assertFalse(old_schema.is_backcompat_with(new_schema))
+        new_schema = parse_schema('any(Tensor self, int dim, bool? keepdim) -> Tensor')
```
This _isn't_ backward compatible. T? doesn't mean that an argument goes unspecified, it means that the caller may pass None or T

houseroad replied:
good catch, fixed.

zdevito commented on test/test_function_schema.py:
```diff
@@ -10,8 +11,81 @@ def test_serialize_and_deserialize(self):
+        self.assertFalse(new_schema.is_backcompat_with(old_schema))
+        self.assertFalse(old_schema.is_backcompat_with(new_schema))
+
+    def test_backward_incompatible_returns(self):
```
Technically it is valid for someone to change a return of Tensor? to Tensor and it is backward compatible.

dzhulgakov replied:
it's not BC is C++ and even in torch script - don't we generate some handling IR around if it's optional? That being said, we don't currently have optional outputs afaiu, so it's fine to be stricter

zdevito replied:
IR is re-generated on load since we are loading from python, so any optional-specific IR won't get generated. This happens in other places as well. For instance, when we inline we are replacing a T? with a T all the time.

houseroad replied:
yes, i added the handling for this case.

zdevito reviewed (changes requested):
Structure looks good, but we should implement correct function subtyping rules.

dzhulgakov reviewed (commented):
And interesting question is handling schemas with "always kwargs". Technically the following two are backward-compatible in Python/JIT because `out` argument is always a kwarg:

`log_sigmoid(Tensor self, *, Tensor(a!) out) -> Tensor(a!)`
`log_sigmoid(Tensor self, int my_new_arg=0, *, Tensor(a!) out) -> Tensor(a!)`

In C++ API we're reordering those kwargs to the front of the signature for outer or backward ops to make sure default arguments are easy to provide at invocation site. So technically it's backward compatible either.

Relevant code:
https://github.com/pytorch/pytorch/blob/33a1c30cb117c2e39126e49b5ec4a2b9b7a64b0d/aten/src/ATen/native_parse.py#L190
https://github.com/pytorch/pytorch/blob/33a1c30cb117c2e39126e49b5ec4a2b9b7a64b0d/tools/jit/gen_jit_dispatch.py#L452

houseroad committed:
Update on "Add isBackwardCompatibleWith for Argument and FunctionSchema"


we intend to be conservative, and will relax the checks in future if necessary.
So far, we consider the following three conditions as backward compatible:
   1) two schemas are equal
   2) two schemas have same number of arguments, and this schema's
      arguments are backward compatible with the corresponding ones in
      argument list of old_schema.
   3) this schema has m argument, old_argument has n argument, m > n.
      the first n arguments of this schema are backward compatible with
      the corresponding arguments of old_schema. the remaning arguments
      must be either OptionalType or provide default values.

Differential Revision: [D16505203](https://our.internmc.facebook.com/intern/diff/D16505203/)

houseroad committed:
Update on "Add isBackwardCompatibleWith for Argument and FunctionSchema"


we intend to be conservative, and will relax the checks in future if necessary.
So far, we consider the following three conditions as backward compatible:
   1) two schemas are equal
   2) two schemas have same number of arguments, and this schema's
      arguments are backward compatible with the corresponding ones in
      argument list of old_schema.
   3) this schema has m argument, old_argument has n argument, m > n.
      the first n arguments of this schema are backward compatible with
      the corresponding arguments of old_schema. the remaning arguments
      must be either OptionalType or provide default values.

Differential Revision: [D16505203](https://our.internmc.facebook.com/intern/diff/D16505203/)

houseroad reviewed (commented):
I have separated the handling for args and kwargs when checking the backward compatibility of function schemas. It should be able to handle the cased mentioned by Dima. :-)

houseroad committed:
Update on "Add isBackwardCompatibleWith for Argument and FunctionSchema"


we intend to be conservative, and will relax the checks in future if necessary.
So far, we consider the following three conditions as backward compatible:
   1) two schemas are equal
   2) two schemas have same number of arguments, and this schema's
      arguments are backward compatible with the corresponding ones in
      argument list of old_schema.
   3) this schema has m argument, old_argument has n argument, m > n.
      the first n arguments of this schema are backward compatible with
      the corresponding arguments of old_schema. the remaning arguments
      must be either OptionalType or provide default values.

Differential Revision: [D16505203](https://our.internmc.facebook.com/intern/diff/D16505203/)

houseroad committed:
Update on "Add isBackwardCompatibleWith for Argument and FunctionSchema"


we intend to be conservative, and will relax the checks in future if necessary.
So far, we consider the following three conditions as backward compatible:
   1) two schemas are equal
   2) two schemas have same number of arguments, and this schema's
      arguments are backward compatible with the corresponding ones in
      argument list of old_schema.
   3) this schema has m argument, old_argument has n argument, m > n.
      the first n arguments of this schema are backward compatible with
      the corresponding arguments of old_schema. the remaning arguments
      must be either OptionalType or provide default values.

Differential Revision: [D16505203](https://our.internmc.facebook.com/intern/diff/D16505203/)

houseroad committed:
Update on "Add isBackwardCompatibleWith for Argument and FunctionSchema"


we intend to be conservative, and will relax the checks in future if necessary.
So far, we consider the following three conditions as backward compatible:
   1) two schemas are equal
   2) two schemas have same number of arguments, and this schema's
      arguments are backward compatible with the corresponding ones in
      argument list of old_schema.
   3) this schema has m argument, old_argument has n argument, m > n.
      the first n arguments of this schema are backward compatible with
      the corresponding arguments of old_schema. the remaning arguments
      must be either OptionalType or provide default values.

Differential Revision: [D16505203](https://our.internmc.facebook.com/intern/diff/D16505203/)

dzhulgakov commented on aten/src/ATen/core/function_schema.h:
```diff
@@ -78,6 +95,18 @@ struct Argument {
+  bool isBackwardCompatibleWith(
+      const Argument& old,
+      bool is_return,
+      std::ostream* why_not) const;
```
nit - make why_not=nullptr too?

dzhulgakov commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -74,6 +105,80 @@ inline std::string FunctionSchema::formatTypeMismatchMsg(
+    std::ostream* why_not) const {
+  if (!(name() == old.name()
+        && overload_name() == old.overload_name()
+        && is_vararg() == old.is_vararg()
```
I'd add a comment here that technically adding vararg is backward compat, but it's internal only, so it'd fine to ignore it for now

houseroad replied:
added

dzhulgakov reviewed (approved):
Great, logic looks correct now!

houseroad committed:
Update on "Add isBackwardCompatibleWith for Argument and FunctionSchema"


we intend to be conservative, and will relax the checks in future if necessary.
So far, we consider the following three conditions as backward compatible:
   1) two schemas are equal
   2) two schemas have same number of arguments, and this schema's
      arguments are backward compatible with the corresponding ones in
      argument list of old_schema.
   3) this schema has m argument, old_argument has n argument, m > n.
      the first n arguments of this schema are backward compatible with
      the corresponding arguments of old_schema. the remaning arguments
      must be either OptionalType or provide default values.

Differential Revision: [D16505203](https://our.internmc.facebook.com/intern/diff/D16505203/)

houseroad commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -51,6 +51,37 @@ inline std::ostream& operator<<(std::ostream& out, const FunctionSchema& schema)
+    if (!rhs->type()->isSubtypeOfExt(lhs->type(), why_not)) {
+      return false;
+    }
+    if (!(detail::defaultValueEquals_(
```
Good suggestion!

zdevito commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -51,6 +51,35 @@ inline std::ostream& operator<<(std::ostream& out, const FunctionSchema& schema)
+    const Argument* lhs = this;
+    const Argument* rhs = &old;
+    // functions are covariant in arguments but contravariant in returns
+    if (is_return) {
```
I looked and _every call site has this as a constant_, so this can simply be accomplished by just swapping the order of the call for the returns. In face this comment makes much more sense inside the return handling then here.

houseroad replied:
Sure, I was debating with myself, whether I should keep the is_return in this function. No strong preference. Updated.

zdevito commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -74,6 +103,82 @@ inline std::string FunctionSchema::formatTypeMismatchMsg(
+    return false;
+  }
+  for (size_t i = 0; i < returns().size(); ++i) {
+    if (!returns().at(i).isBackwardCompatibleWith(
```
Just swap the arguments here and put the contravariant comment here.

houseroad replied:
done

zdevito commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -74,6 +103,82 @@ inline std::string FunctionSchema::formatTypeMismatchMsg(
+      }
+    }
+  }
+  for (auto& kv : old_kwargs) {
```
This is non-deterministic iteration over an unordered map. While it shouldn't change the result it can make debugging harder since it any bugs in this pathway will inherit the non-determinism.

houseroad replied:
Changed to map, the number of args should be small, so I think it's acceptable.

zdevito commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -74,6 +103,82 @@ inline std::string FunctionSchema::formatTypeMismatchMsg(
+      return false;
+    }
+  }
+  for (size_t i = old_args.size(); i < args.size(); ++i) {
```
Please add a comment about what this is doing.

houseroad replied:
added

zdevito commented on aten/src/ATen/core/function_schema_inl.h:
```diff
@@ -74,6 +103,82 @@ inline std::string FunctionSchema::formatTypeMismatchMsg(
+    }
+    kwargs.erase(it);
+  }
+  for (auto& kv : kwargs) {
```
More comments here, I can no longer follow the logic.

houseroad replied:
sure

zdevito reviewed (approved):
Looks good. We should remove the is_return bool since it is not needed. There are a few places that need comments explaining the valid migrations from kwarg_only to normal arguments. I am not sure that I follow the logic, so I can't say whether it is correct or not. I also am not sure it matters, have there been cases where we moved a kwarg_only arg to be a normal arg?

houseroad committed:
Update on "Add isBackwardCompatibleWith for Argument and FunctionSchema"


we intend to be conservative, and will relax the checks in future if necessary.
So far, we consider the following three conditions as backward compatible:
   1) two schemas are equal
   2) two schemas have same number of arguments, and this schema's
      arguments are backward compatible with the corresponding ones in
      argument list of old_schema.
   3) this schema has m argument, old_argument has n argument, m > n.
      the first n arguments of this schema are backward compatible with
      the corresponding arguments of old_schema. the remaning arguments
      must be either OptionalType or provide default values.

Differential Revision: [D16505203](https://our.internmc.facebook.com/intern/diff/D16505203/)

houseroad committed:
Update on "Add isBackwardCompatibleWith for Argument and FunctionSchema"


we intend to be conservative, and will relax the checks in future if necessary.
So far, we consider the following three conditions as backward compatible:
   1) two schemas are equal
   2) two schemas have same number of arguments, and this schema's
      arguments are backward compatible with the corresponding ones in
      argument list of old_schema.
   3) this schema has m argument, old_argument has n argument, m > n.
      the first n arguments of this schema are backward compatible with
      the corresponding arguments of old_schema. the remaning arguments
      must be either OptionalType or provide default values.

Differential Revision: [D16505203](https://our.internmc.facebook.com/intern/diff/D16505203/)

houseroad commented:
@pytorchbot retest this please
